File: logic/dsu.py
"""Disjoint-set union (union-find) over vertices, tracking special-subset
membership counts per component.
"""
from typing import Any, Hashable


class DSU:
    """Union-find structure over allVertices, tracking how many members of
    specialSubset currently belong to each component.
    """

    def __init__(self, allVertices: list, specialSubset: list) -> None:
        self.specialSubset: list = list(specialSubset)
        specialAsSet = set(self.specialSubset)

        self._parent: dict[Hashable, Hashable] = {vertex: vertex for vertex in allVertices}
        self._size: dict[Hashable, int] = {vertex: 1 for vertex in allVertices}
        self._specialCount: dict[Hashable, int] = {
            vertex: (1 if vertex in specialAsSet else 0) for vertex in allVertices
        }
        self._componentCount: int = len(allVertices)

    def find(self, vertex: Any) -> Any:
        root = vertex
        while self._parent[root] != root:
            root = self._parent[root]

        # Path compression.
        current = vertex
        while self._parent[current] != root:
            nextVertex = self._parent[current]
            self._parent[current] = root
            current = nextVertex

        return root

    def union(self, a: Any, b: Any) -> Any:
        rootA = self.find(a)
        rootB = self.find(b)
        if rootA == rootB:
            return rootA

        smallerRoot, largerRoot = sorted(
            (rootA, rootB), key=lambda root: self._size[root]
        )

        self._parent[smallerRoot] = largerRoot
        self._size[largerRoot] += self._size[smallerRoot]
        self._specialCount[largerRoot] += self._specialCount[smallerRoot]
        del self._size[smallerRoot]
        del self._specialCount[smallerRoot]
        self._componentCount -= 1

        return largerRoot

    def sCount(self, root: Any) -> int:
        return self._specialCount.get(root, 0)

    def componentCount(self) -> int:
        return self._componentCount

File: logic/dsu.js
/**
 * Disjoint-set union (union-find) over vertices, tracking special-subset
 * membership per vertex.
 */
import { vertexKey } from "./vertices.js";

// PUBLIC INTERFACE

export class DSU {
  constructor(allVertices, specialSubset) {
    this._specialKeys = new Set(specialSubset.map(vertexKey));

    this._representative = new Map();
    this._parent = new Map();
    this._size = new Map();

    for (const vertex of allVertices) {
      const key = vertexKey(vertex);
      this._representative.set(key, vertex);
      this._parent.set(key, key);
      this._size.set(key, 1);
    }

    this._componentCount = allVertices.length;
  }

  find(vertex) {
    const startKey = vertexKey(vertex);
    let rootKey = startKey;
    while (this._parent.get(rootKey) !== rootKey) {
      rootKey = this._parent.get(rootKey);
    }
    this._compressPath(startKey, rootKey);
    return this._representative.get(rootKey);
  }

  _compressPath(startKey, rootKey) {
    let currentKey = startKey;
    while (this._parent.get(currentKey) !== rootKey) {
      const nextKey = this._parent.get(currentKey);
      this._parent.set(currentKey, rootKey);
      currentKey = nextKey;
    }
  }

  union(a, b) {
    const rootA = this.find(a);
    const rootB = this.find(b);
    const rootAKey = vertexKey(rootA);
    const rootBKey = vertexKey(rootB);
    if (rootAKey === rootBKey) {
      return rootA;
    }

    const [smallerKey, largerKey] =
      this._size.get(rootAKey) <= this._size.get(rootBKey) ? [rootAKey, rootBKey] : [rootBKey, rootAKey];

    this._parent.set(smallerKey, largerKey);
    this._size.set(largerKey, this._size.get(largerKey) + this._size.get(smallerKey));
    this._size.delete(smallerKey);
    this._componentCount -= 1;
    return this._representative.get(largerKey);
  }

  componentSize(vertex) {
    return this._size.get(vertexKey(this.find(vertex))) ?? 0;
  }

  componentCount() {
    return this._componentCount;
  }

  isSpecial(vertex) {
    return this._specialKeys.has(vertexKey(vertex));
  }
}
